liblangutil/ParserBase.cpp

```diff
@@ -57,30 +77,80 @@ Token ParserBase::advance()
 	return m_scanner->next();
 }
 
+std::string ParserBase::tokenName(Token _token)
+{
+	if (_token == Token::Identifier)
+		return string("identifier");
+	else if (_token == Token::EOS)
+		return string("end of source");
+	else if (TokenTraits::isReservedKeyword(_token))
+		return string("reserved keyword '") + TokenTraits::friendlyName(_token) + "'";
+	else if (TokenTraits::isElementaryTypeName(_token)) //for the sake of accuracy in reporting
+	{
+		ElementaryTypeNameToken elemTypeName = m_scanner->currentElementaryTypeNameToken();
+		return string("'") + elemTypeName.toString() + "'";
+	}
+	else
+		return string("'") + TokenTraits::friendlyName(_token) + "'";
+};
+
 void ParserBase::expectToken(Token _value, bool _advance)
 {
 	Token tok = m_scanner->currentToken();
 	if (tok != _value)
 	{
-		auto tokenName = [this](Token _token)
+		std::string const expectToken = ParserBase::tokenName(_value);
+		parserError(string("Expected ") + expectToken + string(" but got ") + tokenName(tok));
+		// Do not advance so that recovery can sync or make use of the current token. This is especially useful if the expected token
+		// is the only one that is missing and is at the end of a construct.
+		// "{ ... ; }" is such an example.
+		//        ^
+		_advance = false;
+	}
+	if (_advance)
+		m_scanner->next();
+}
+
+void ParserBase::expectTokenOrConsumeUntil(Token _value, char const *_lhs, bool _advance)
+{
+	Token tok = m_scanner->currentToken();
+	if (tok != _value)
+	{
+		Token token = m_scanner->currentToken();
+		int startPosition = position();
+		SourceLocation errorLoc = SourceLocation{startPosition, endPosition(), source()};
+		while (token != _value && token != Token::EOS)
+			token = m_scanner->next();
+		std::string const expectToken = ParserBase::tokenName(_value);
+		std::string const mess = fmt::sprintf("In <%s>, %s is expected; got %s instead.", _lhs, expectToken.c_str(), ParserBase::tokenName(tok).c_str());
+		if (token == Token::EOS)
 		{
-			if (_token == Token::Identifier)
-				return string("identifier");
-			else if (_token == Token::EOS)
-				return string("end of source");
-			else if (TokenTraits::isReservedKeyword(_token))
-				return string("reserved keyword '") + TokenTraits::friendlyName(_token) + "'";
-			else if (TokenTraits::isElementaryTypeName(_token)) //for the sake of accuracy in reporting
-			{
-				ElementaryTypeNameToken elemTypeName = m_scanner->currentElementaryTypeNameToken();
-				return string("'") + elemTypeName.toString() + "'";
-			}
+			// rollback to where the token started, and raise exception to be caught at a higher level.
```

Comment:
Why not go to EOS?

Reply:
Because _recovery_ is what we want, so that an AST can be produced with more information. 

It feels to me like a lot of the focus is on errors for batch compile users, and that is a lot of the use right now. 

It feels to me that we need two modes, one where there is a lot of detail given so that an IDE can use it, and one for batch use. What do you think?